### PR TITLE
QA-273: Feature/python test runner

### PR DESCRIPTION
**Pull request by dothebart**
replace fish and ps test launching facilities plus report generators for one python implementation
 - the environment variable TESTSUITE_TIMEOUT defines a deadline to the tests, how much seconds should be allowed.
 - tests are running in worker threads.
 - main thread keeps control, launches more worker threads, once machine bandwith permits, but only every 5s as closest.
 - tests themselves have their timeouts; testing.js will abort if they are reached.
 - workers have a progressive timeout, if it doesn't hear back from testing.js for 999999999s it will hard kill and abort.
 - if workers have no output from testing.js they check whether the deadline is reached.
 - if the deadline is reached, SIG_INT[*nix] / SIG_BREAK[windows] is sent to testing.js to trigger its deadline feature.
 - the reached deadline will be indicated to testfailures.txt and the logfile of this test.
 - with deadline engageged, testing.js can send no more subsequent requests, nor spawn processes => eventually testing will abort.
 - force shutdown of instances will reset the deadline, SIG_ABRT arangods, and try to do core dump analysis.
 - workers continue reading pipes from testing.js, but once no chars are comming, waitpid() checks with a 1s timout whether testing.js is done.
 - if the worker reaches 180 counters of waitpid() it will give up. It will hard kill testing.js and all other child processes.
 - this should unblock the workers STDOUT/STDERR threads, and they should exit.
 - the waitpid() on testing.js should exit, I/O threads should be joined, results should be passed up to the main thread.
 - so the workers still have a slugish interpretation of the deadline, giving them the chance to collect as much knowledge as posible.
 - meanwhile the main thread has a *fixed* deadline: 5 minutes after the TESTSUITE_TIMEOUT is reached.
 - if not all workers have indicated their exit before this final deadline:
 - the main thread will start killing any subprocesses of itself which it finds.
 - after this wait another 20s, to see whether the workers may have been unblocked by the killing
 - if not, it shouts "Geronimoooo" and takes the big shotgun, and force-terminates the python process which is running it. This will kill all threads as well and terminate the process.
 - if all workers have indicated their exited in time, their threads will be joined.
 - reports will be generated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -234,6 +234,100 @@ if supported, set number of concurrent builds to `PARALLELISM`
 
 ## Testing
 
+`jenkins/helper/test_launch_controller.py` is used to control multiple test executions.
+
+### Its dependencies over stock python3 are:
+ - psutil to controll subprocesses
```
Comment (jsteemann) [inline README.md:240]:
```suggestion
 - psutil to control subprocesses
```

File: README.md [hunk 2]
```diff
@@ -234,6 +234,100 @@ if supported, set number of concurrent builds to `PARALLELISM`
...
+ - py7zr (optional) to build 7z reports instead of tar.bz2
+
+### Its reading these environment variables:
```
Comment (jsteemann) [inline README.md:243]:
```suggestion
### It's reading these environment variables:
```

File: README.md [hunk 3]
```diff
@@ -234,6 +234,100 @@ if supported, set number of concurrent builds to `PARALLELISM`
...
+ - `PATH/test-definition.txt` - (first parameter) test definitions file from the arangodb source tree
+   (also used to locate the arangodb source)
+ - `-f` `[launch|dump]` use `dump` for syntax checking of `test-definition.txt` instead of executing the tests
+ - `--validate-only` don't run the tests
+ - `--help-flags` list the flags which can be used in `test-definition.txt`:
+    - `cluster`: this test requires a cluster
+    - `single`: this test requires a single server
+    - `full`: this test is only executed in full tests
+    - `!full`: this test is only executed in non-full tests
+    - `gtest`: only the gtest are to be executed
```
Comment (jsteemann) [inline README.md:272]:
```suggestion
    - `gtest`: only the gtests are to be executed
```

File: README.md [hunk 4]
```diff
@@ -234,6 +234,100 @@ if supported, set number of concurrent builds to `PARALLELISM`
...
+    - `ldap`: ldap
+    - `enterprise`: this tests is only executed with the enterprise version
```
Comment (jsteemann) [inline README.md:274]:
```suggestion
    - `enterprise`: this test is only executed with the enterprise version
```

File: README.md [hunk 5]
```diff
@@ -234,6 +234,100 @@ if supported, set number of concurrent builds to `PARALLELISM`
...
+    - `!windows`: test is excluded from ps1 output
+ - `--cluster` filter `test-definition.txt` for all tests flagged as `cluster`
+ - `--full` - all tests including those flagged as `full` are executed.
+ - `--gtest` - only gtest are executed
```
Comment (jsteemann) [inline README.md:278]:
```suggestion
 - `--gtest` - only gtests are executed
```

**Review by jsteemann — APPROVED** [review]
Non-Python files LGTM.
I am not in a good position to review the Python code in this PR though.

**Review by fceller — APPROVED** [review]

**Pull request merged**